File: verify_gemini_key.js
const https = require('https');

// Colors for console output
const colors = {
    reset: "\x1b[0m",
    red: "\x1b[31m",
    green: "\x1b[32m",
    yellow: "\x1b[33m",
    cyan: "\x1b[36m",
    bold: "\x1b[1m"
};

/**
 * 驗證 Gemini API Key 的獨立腳本
 * 使用 Node.js 原生 modules，無需 npm install
 */
async function verifyGeminiKey(apiKey) {
    console.clear();
    console.log(`${colors.cyan}${colors.bold}════════════════════════════════════════════${colors.reset}`);
    console.log(`${colors.cyan}   🔑 Gemini API Key Validator${colors.reset}`);
    console.log(`${colors.cyan}════════════════════════════════════════════${colors.reset}\n`);

    if (!apiKey) {
        console.error(`${colors.red}❌ 未提供 API Key${colors.reset}`);
        console.log(`${colors.yellow}請使用以下方式執行:${colors.reset}`);
        console.log(`   node verify_gemini_key.js ${colors.bold}<YOUR_API_KEY>${colors.reset}`);
        return;
    }

    // 嘗試使用較新的模型
    const model = 'gemini-2.5-flash';
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;

    console.log(`📡 正在連線至 Google Gemini API...`);
    console.log(`⚙️ 模型: ${colors.green}${model}${colors.reset}`);
    console.log(`🔑 Key : ${apiKey.substring(0, 8)}...${apiKey.substring(apiKey.length - 4)}\n`);

    const payload = {
        contents: [{
            parts: [{ text: "Reply with 'Valid' if you can read this." }]
        }],
        generationConfig: {
            maxOutputTokens: 20
        }
    };

    const startTime = Date.now();

    const options = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' }
    };

    const req = https.request(url, options, (res) => {
        let data = '';

        res.on('data', chunk => data += chunk);

        res.on('end', () => {
            const result = parseResponse(data, res.statusCode);
            const duration = ((Date.now() - startTime) / 1000).toFixed(2);

            if (result.success) {
                console.log(`${colors.green}✅ 驗證成功 (Success)${colors.reset}`);
                console.log(`⏱️ 耗時: ${duration}s`);
                console.log(`🤖 回應: ${result.message}`);
            } else {
                console.log(`${colors.red}❌ 驗證失敗 (Failed)${colors.reset}`);
                console.log(`⏱️ 耗時: ${duration}s`);
                console.log(`⚠️ 原因: ${result.error}`);

                if (res.statusCode === 400) console.log(`${colors.yellow}hint: 檢查 Key 是否格式正確或包含無效字符${colors.reset}`);
                if (res.statusCode === 401) console.log(`${colors.yellow}hint: Key 無效或已過期${colors.reset}`);
                if (res.statusCode === 429) console.log(`${colors.yellow}hint: 達到配額限制 (Quota Exceeded)${colors.reset}`);
            }
            console.log(`\n${colors.cyan}════════════════════════════════════════════${colors.reset}`);
        });
    });

    req.on('error', (e) => {
        console.error(`\n${colors.red}❌ 連線錯誤:${colors.reset} ${e.message}`);
    });

    req.write(JSON.stringify(payload));
    req.end();
}

function parseResponse(data, status) {
    try {
        const json = JSON.parse(data);
        if (status >= 200 && status < 300) {
            const text = json.candidates?.[0]?.content?.parts?.[0]?.text?.trim();
            return { success: true, message: text || '(No content)' };
        } else {
            const msg = json.error?.message || json.error?.status || 'Unknown error';
            return { success: false, error: `${status} ${msg}` };
        }
    } catch (e) {
        return { success: false, error: `${status} Invalid JSON response: ${data.substring(0, 100)}...` };
    }
}

// Get key from args
const argKey = process.argv[2];
verifyGeminiKey(argKey);

File: verify_gemini_key.py
import sys
import json
import requests
import time
import io

# Force UTF-8 Output for Windows Terminals
if sys.platform == 'win32':
    sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='backslashreplace')
    sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding='utf-8', errors='backslashreplace')

# Colors for console output
class Colors:
    RESET = "\033[0m"
    RED = "\033[31m"
    GREEN = "\033[32m"
    YELLOW = "\033[33m"
    CYAN = "\033[36m"
    BOLD = "\033[1m"

def verify_gemini_key(api_key):
    print(f"\n{Colors.CYAN}{Colors.BOLD}════════════════════════════════════════════{Colors.RESET}")
    print(f"{Colors.CYAN}   🔑 Gemini API Key Validator (Python Edition){Colors.RESET}")
    print(f"{Colors.CYAN}════════════════════════════════════════════{Colors.RESET}\n")

    if not api_key:
        print(f"{Colors.RED}❌ ERROR: API Key not provided{Colors.RESET}")
        print(f"{Colors.YELLOW}Usage:{Colors.RESET}")
        print(f"   python verify_gemini_key.py {Colors.BOLD}<YOUR_API_KEY>{Colors.RESET}")
        return

    model = 'gemini-2.5-flash'
    url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={api_key}"

    print(f"📡 Connecting to Google Gemini API...")
    print(f"⚙️ Model: {Colors.GREEN}{model}{Colors.RESET}")
    
    if len(api_key) > 12:
        masked_key = f"{api_key[:8]}...{api_key[-4:]}"
    else:
        masked_key = "***"
    print(f"🔑 Key : {masked_key}\n")

    payload = {
        "contents": [{
            "parts": [{"text": "Reply with 'Valid' if you can read this."}]
        }],
        "generationConfig": {
            "maxOutputTokens": 20
        }
    }

    start_time = time.time()

    try:
        response = requests.post(url, json=payload, headers={'Content-Type': 'application/json'}, timeout=10)
        duration = time.time() - start_time
        
        # Check HTTP status code
        if response.status_code == 200:
            try:
                data = response.json()
                candidates = data.get('candidates', [])
                if candidates:
                    parts = candidates[0].get('content', {}).get('parts', [])
                    if parts:
                        description = parts[0].get('text', '(No text content)').strip()
                    else:
                        description = "(No parts in content)"
                else:
                    description = "(No candidates returned)"
                
                print(f"{Colors.GREEN}✅ Validation Success!{Colors.RESET}")
                print(f"⏱️ Time: {duration:.2f}s")
                print(f"🤖 Response: {description}")
            except Exception as parse_err:
                print(f"{Colors.YELLOW}⚠️ Valid HTTP 200 but failed to parse response: {parse_err}{Colors.RESET}")
                print(f"Raw: {response.text[:100]}...")

        else:
            print(f"{Colors.RED}❌ Validation Failed!{Colors.RESET}")
            print(f"⏱️ Time: {duration:.2f}s")
            print(f"⚠️ Status: {response.status_code}")
            
            try:
                error_data = response.json()
                error_msg = error_data.get('error', {}).get('message', response.text)
                print(f"⚠️ Reason: {error_msg}")
                
                if response.status_code == 400:
                    print(f"{Colors.YELLOW}hint: Check if Key format is correct.{Colors.RESET}")
                elif response.status_code == 401:
                    print(f"{Colors.YELLOW}hint: Key is invalid or expired.{Colors.RESET}")
                elif response.status_code == 404:
                     print(f"{Colors.YELLOW}hint: Model '{model}' not found. Try a different model.{Colors.RESET}")
            except:
                print(f"⚠️ Response: {response.text}")

    except Exception as e:
        print(f"\n{Colors.RED}❌ Connection Error:{Colors.RESET} {str(e)}")
        print(f"{Colors.YELLOW}Check your internet connection or proxy settings.{Colors.RESET}")

    print(f"\n{Colors.CYAN}════════════════════════════════════════════{Colors.RESET}")

if __name__ == "__main__":
    if len(sys.argv) > 1:
        verify_gemini_key(sys.argv[1])
    else:
        # Prompt user if no arg provided
        print("Please enter your Gemini API Key:")
        try:
            key = input("> ").strip()
            verify_gemini_key(key)
        except EOFError:
            pass
